### ISSUE TITLE
Source Trained Models

### ISSUE THREAD

**Issue by Sairam13001** [junia3/LPLD#6]
Hi @junia3 

Could you please let me know how I can access source trained models.

**Comment by Re-dot-art** [comment]
> Hi [@junia3](https://github.com/junia3)
> 
> Could you please let me know how I can access source trained models.

May I ask if you have found the pre training weights ？

**Comment by junia3** [comment]
You can obtain the source-trained model from the official repository of IRG-SFDA (CVPR 2023) at: https://github.com/Vibashan/irg-sfda.

Alternatively, you can pretrain the model on the source-domain dataset, such as Cityscapes, KITTI, Sim, or VOC (Recommended).

For cases where pretrained model weights are unavailable, we manually trained the corresponding models (ResNet-50, VGG) using the source datasets.

**Comment by Sairam13001** [comment]
Hi @junia3, thank you for your reply.

I took the source-trained weights for the VOC dataset from the IRG-SFDA repo (https://github.com/Vibashan/irg-sfda) and ran the train_main.py for 10 epochs for the VOC2Clipart adaptation. I ran the training with different seeds and the best AP50 the teacher could reach is only 32.46 (that too in the 7th epoch).  I am attaching the log file for your reference. Note that, this file has 20 AP50 values, which are basically the student_model AP50 and teacher_model AP50 after each of the 10 epochs. 

Please let me know if I am doing something wrong. Why can't I reach the results shown in the paper?

[log.txt](https://github.com/user-attachments/files/19073221/log.txt)

**Comment by Sairam13001** [comment]
I am facing the same problem for City2Foggy also. I tried with multiple seeds but the performance does not reach what is given in the paper (in any of the epochs). Please let me know if I am missing something.

**Comment by junia3** [comment]
It seems that there is a bug in the source code
I will fix it and upload it again. @Sairam13001

**Comment by Sairam13001** [comment]
Sure, thank you for your prompt reply. I will be waiting for your fix.

**Comment by Sairam13001** [comment]
@kaustubhrb 

I did not try the Kitti2city setting yet, however, I was not able to get the reported numbers for the VOC2Clipart and Cityscapes2Foggy settings. 

But I think in the Kitti2City config, the num_classes is given as 2 although there is only one class (car). So, you might have to multiply the AP50 you got by 2, which makes it 25.7461*2 = 51.4922

**Comment by Sairam13001** [comment]
@kaustubhrb

Did you train on the target domain starting from the source domain weights, or did you directly evaluate using the target-trained weights?

**Comment by kaustubhrb** [comment]
No, this is just inferencing using the given best weights on the repo for this setting.

**Comment by Sairam13001** [comment]
I just reviewed the code again, and here’s what I observed:

In the Sim10k2City config, num_classes is correctly set to 1, whereas in the Kitti2City config, it is set to 2, even though both settings have only one foreground class—Car.

In detectron2/modeling/roi_heads/fast_rcnn.py, at line 224, the classifier is defined as:
self.cls_score = nn.Linear(input_size, num_classes + 1)
This means the model predicts logits for (foreground classes + 1 extra for background).
However, in the Kitti2City setup, since num_classes is 2, the classifier ends up predicting logits for three classes, mistakenly treating Car and Background as foreground and adding an extra class for the background again. 

Correct me if I am missing something here.

**Comment by kaustubhrb** [comment]
Thanks for the suggestion. I tried this but unfortunately the given model weights for K2C has 3 classes (2+1) defined in the final classifier layer, hence while loading the weights it skips the final layer due to incompatible tensor dim (with 2 classes now (1+1)). Here is the warning message. 

![Image](https://github.com/user-attachments/assets/6ed818f3-e5c7-4c23-9bfd-4b97cfa2d19e)


Lastly the resulting scores are even lesser than before when num_classes is set to 1 like S2C. 

![Image](https://github.com/user-attachments/assets/dd81fa14-f9bf-4f04-b82b-04f9477036a6)

**Comment by Sairam13001** [comment]
I did not mean to suggest changing the num_classes to 1. I think we should only consider the AP of the car class in the case of Kitti2City instead of the mean AP across classes.

**Comment by kaustubhrb** [comment]
I see. But I'm not fully sure about multiplying the AP with 2 here. Thanks for your input though. 

I hope @junia3 can clarify this. If there is some issue with the number of classes for a certain setting, it's better if the code is updated with the correct version rather than making than us making this change on our own.
